pkg/reconciler/brokercell/brokercell.go

```diff
@@ -43,57 +44,85 @@ type envConfig struct {
 	MetricsPort        int    `envconfig:"METRICS_PORT" default:"9090"`
 }
 
+// NewReconciler creates a new BrokerCell reconciler.
+func NewReconciler(base *reconciler.Base, serviceLister corev1listers.ServiceLister, endpointsLister corev1listers.EndpointsLister, deploymentLister appsv1listers.DeploymentLister) (*Reconciler, error) {
+	var env envConfig
+	if err := envconfig.Process("BROKER_CELL", &env); err != nil {
+		return nil, err
+	}
+	svcRec := &reconciler.ServiceReconciler{
+		KubeClient:      base.KubeClientSet,
+		ServiceLister:   serviceLister,
+		EndpointsLister: endpointsLister,
+		Recorder:        base.Recorder,
+	}
+	deploymentRec := &reconciler.DeploymentReconciler{
+		KubeClient: base.KubeClientSet,
+		Lister:     deploymentLister,
+		Recorder:   base.Recorder,
+	}
+	r := &Reconciler{
+		Base:             base,
+		serviceLister:    serviceLister,
+		endpointsLister:  endpointsLister,
+		deploymentLister: deploymentLister,
+		env:              env,
+		svcRec:           svcRec,
+		deploymentRec:    deploymentRec,
+	}
+	return r, nil
+}
+
 // Reconciler implements controller.Reconciler for BrokerCell resources.
 type Reconciler struct {
 	*reconciler.Base
 
-	kubeClientSet    kubernetes.Interface
-	brokerLister     eventinglisters.BrokerLister
 	serviceLister    corev1listers.ServiceLister
 	endpointsLister  corev1listers.EndpointsLister
 	deploymentLister appsv1listers.DeploymentLister
 
+	svcRec        *reconciler.ServiceReconciler
+	deploymentRec *reconciler.DeploymentReconciler
+
 	env envConfig
 }
 
 // Check that our Reconciler implements Interface
 var _ bcreconciler.Interface = (*Reconciler)(nil)
 
-// Optionally check that our Reconciler implements Finalizer
-var _ bcreconciler.Finalizer = (*Reconciler)(nil)
-
 // ReconcileKind implements Interface.ReconcileKind.
 func (r *Reconciler) ReconcileKind(ctx context.Context, bc *intv1alpha1.BrokerCell) pkgreconciler.Event {
 	bc.Status.InitializeConditions()
 
 	// Reconcile ingress deployment and service
 	ingressArgs := r.makeIngressArgs(bc)
-	if _, err := reconciler.ReconcileDeployment(r.kubeClientSet, r.deploymentLister, resources.MakeIngressDeployment(ingressArgs)); err != nil {
-		r.Logger.Errorf("Failed to reconcile ingress deployment for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
-		bc.Status.MarkIngressFailed("IngressDeploymentFailed", "Failed to reconcile ingress deployment for \"%s/%s\": %v", bc.Namespace, bc.Name, err)
+	if _, err := r.deploymentRec.ReconcileDeployment(bc, resources.MakeIngressDeployment(ingressArgs)); err != nil {
+		r.Logger.Desugar().Error("Failed to reconcile ingress deployment for \"%s/%s\": %v", zap.Any("namespace", bc.Namespace), zap.Any("name", bc.Name), zap.Error(err))
```

Comment:
Sorry I misread this before and that made my earlier comment unclear. Instead of using `r.Logger`, get the logger from the context using `logging.FromContext(ctx)`. It contains additional structured fields specific to this reconcile call. You can use `knative.dev/eventing/pkg/logging` (which returns a desugared logger) to eliminate the `Desugar()` calls.